File: ftmq/query/rql.py
"""
[RQL](https://github.com/pjwerneck/pyrql) (Resource Query Language) bridge.

RQL is a URL-friendly query language of nestable named operators - e.g.
`and(eq(schema,Person),or(eq(properties.name,jane),eq(countries,de)))` - which
maps directly onto the ftmq `Expr` tree (`and`/`or`/`not` + comparison leaves).
Unlike the flat Aleph param grammar, RQL expresses arbitrary nesting, so this is
the way to carry a full `M & (P | G)` tree through a single string.

Field names follow the Aleph convention (`schema`, `dataset`, `properties.<name>`,
a `registry.groups` name, or `origin`); a bare name that matches none of those is
treated as an FtM property.
"""

from __future__ import annotations

from typing import Any

import pyrql  # type: ignore[import-untyped]

from ftmq.query.aleph import _FAMILIES, _aleph_field, _resolve_field
from ftmq.query.exceptions import QueryError
from ftmq.query.leaves import Leaf
from ftmq.query.nodes import AND, OR, Expr, combine

# RQL comparison operator -> ftmq comparator
RQL_COMPARATORS = {
    "eq": "eq",
    "ne": "not",
    "lt": "lt",
    "le": "lte",
    "gt": "gt",
    "ge": "gte",
    "in": "in",
    "out": "not_in",
    "like": "like",
    "ilike": "ilike",
    "contains": "like",
}

# ftmq comparator -> RQL operator (the expressible subset; `null`, `startswith`,
# `endswith`, `notlike`, `notilike`, `between` have no RQL equivalent)
TO_RQL_OPERATORS = {
    "eq": "eq",
    "not": "ne",
    "lt": "lt",
    "lte": "le",
    "gt": "gt",
    "gte": "ge",
    "in": "in",
    "not_in": "out",
    "like": "like",
    "ilike": "ilike",
}


def _resolve_rql_field(field: str) -> tuple[str, str]:
    try:
        return _resolve_field(field)
    except QueryError:
        # a bare name that is not meta / group / context / `properties.` is
        # treated as a property; validity is checked when the leaf is built
        return "P", field


def _rql_leaf(op: str, args: list[Any]) -> Expr:
    comparator = RQL_COMPARATORS.get(op)
    if comparator is None:
        raise QueryError(f"Unsupported RQL operator: `{op}`")
    field, value = args[0], args[1]
    family, key = _resolve_rql_field(field)
    if comparator in ("in", "not_in"):
        value = list(value)
    elif comparator in ("like", "ilike") and isinstance(value, str):
        # RQL uses `*` as the wildcard; ftmq `like`/`ilike` is substring-based
        value = value.replace("*", "")
    lookup = key if comparator == "eq" else f"{key}__{comparator}"
    return _FAMILIES[family](**{lookup: value})


def rql_to_expr(data: dict[str, Any]) -> Expr:
    """Convert a parsed RQL AST (`{"name": ..., "args": [...]}`) to an `Expr`."""
    op, args = data["name"], data["args"]
    if op == "and":
        result = combine(*(rql_to_expr(a) for a in args), connector=AND)
    elif op == "or":
        result = combine(*(rql_to_expr(a) for a in args), connector=OR)
    elif op == "not":
        return ~rql_to_expr(args[0])
    else:
        return _rql_leaf(op, args)
    if result is None:
        raise QueryError(f"Empty RQL group: `{op}`")
    return result


def parse_rql(value: str) -> Expr | None:
    """Parse an RQL query string into an `Expr` (or `None` if empty).

    Raises:
        QueryError: If the RQL uses an unsupported operator or field.
    """
    data = pyrql.parse(value)
    return rql_to_expr(data) if data else None


def _leaf_to_rql(leaf: Leaf) -> dict[str, Any]:
    op = TO_RQL_OPERATORS.get(str(leaf.comparator))
    if op is None:
        raise QueryError(f"Comparator `{leaf.comparator}` is not expressible as RQL")
    value = leaf.value
    if op in ("in", "out"):
        value = tuple(sorted(str(v) for v in value))
    return {"name": op, "args": [_aleph_field(leaf), value]}


def expr_to_rql(expr: Expr) -> dict[str, Any]:
    """Convert an `Expr` tree to an RQL AST (`{"name": ..., "args": [...]}`)."""
    group = "or" if expr.connector == OR else "and"
    parts: list[dict[str, Any]] = []
    for child in expr.children:
        if isinstance(child, Expr):
            child_ast = expr_to_rql(child)
            # flatten a non-negated same-connector subgroup into this one
            if not child.negated and child_ast.get("name") == group:
                parts.extend(child_ast["args"])
            else:
                parts.append(child_ast)
        else:
            parts.append(_leaf_to_rql(child))
    if not parts:
        raise QueryError("Cannot serialize an empty query to RQL")
    # a single-child group is just that child
    body = parts[0] if len(parts) == 1 else {"name": group, "args": parts}
    if expr.negated:
        return {"name": "not", "args": [body]}
    return body


def to_rql(expr: Expr | None) -> str:
    """Serialize an `Expr` tree to an RQL query string.

    Raises:
        QueryError: If a leaf uses a comparator with no RQL equivalent (`null`,
            `startswith`, `endswith`, ...).
    """
    if expr is None or not expr:
        return ""
    return str(pyrql.unparse(expr_to_rql(expr)))

File: js/query/rql.ts
import { Agg, type AggFunc } from "./aggregations.js";
import { resolveField } from "./aleph.js";
import { QueryError } from "./exceptions.js";
import { type Leaf } from "./leaves.js";
import { AND, combine, Expr, FAMILIES, OR } from "./nodes.js";
import { byString } from "./util.js";

// --- the self-contained RQL codec (pyrql-compatible `{name, args}` AST) ------

export type RqlArg = string | RqlNode | RqlArg[];
export interface RqlNode {
  name: string;
  args: RqlArg[];
}

const isNode = (arg: RqlArg): arg is RqlNode =>
  typeof arg === "object" && !Array.isArray(arg);

/**
 * Parse an RQL string into a `{name, args}` AST. Values are taken verbatim
 * (already url-decoded by the URL layer), matching pyrql's unencoded output; a
 * tuple `(a,b)` is represented as an array.
 */
export function parseRql(input: string): RqlNode | null {
  const text = input.trim();
  if (!text) return null;
  let i = 0;

  const parseArgs = (): RqlArg[] => {
    i++; // consume "("
    const args: RqlArg[] = [];
    if (text[i] === ")") {
      i++;
      return args;
    }
    for (;;) {
      args.push(parseTerm());
      if (text[i] === ",") {
        i++;
        continue;
      }
      if (text[i] === ")") {
        i++;
        break;
      }
      throw new QueryError(`Invalid RQL near: \`${text.slice(i)}\``);
    }
    return args;
  };

  const parseTerm = (): RqlArg => {
    if (text[i] === "(") return parseArgs(); // tuple
    const start = i;
    while (i < text.length && !"(),".includes(text[i])) i++;
    const token = text.slice(start, i);
    if (text[i] === "(") return { name: token, args: parseArgs() };
    return token; // bare value
  };

  const node = parseTerm();
  if (!isNode(node)) {
    throw new QueryError("Invalid RQL: expected an operator");
  }
  return node;
}

/** Serialize a `{name, args}` AST to an RQL string (no encoding, like pyrql). */
export function unparseRql(node: RqlNode): string {
  const arg = (a: RqlArg): string =>
    typeof a === "string"
      ? a
      : Array.isArray(a)
        ? `(${a.map(arg).join(",")})`
        : unparseRql(a);
  return `${node.name}(${node.args.map(arg).join(",")})`;
}

// --- the tree <-> RQL bridge (mirrors ftmq.query.rql) ------------------------

const RQL_COMPARATORS: Record<string, string> = {
  eq: "eq",
  ne: "not",
  lt: "lt",
  le: "lte",
  gt: "gt",
  ge: "gte",
  in: "in",
  out: "not_in",
  like: "like",
  ilike: "ilike",
  contains: "like",
};
const TO_RQL_OPERATORS: Record<string, string> = {
  eq: "eq",
  not: "ne",
  lt: "lt",
  lte: "le",
  gt: "gt",
  gte: "ge",
  in: "in",
  not_in: "out",
  like: "like",
  ilike: "ilike",
};
const RQL_FUNCTIONS: Record<string, AggFunc> = {
  sum: "sum",
  min: "min",
  max: "max",
  mean: "avg",
  count: "count",
};
const TO_RQL_FUNCTIONS: Record<string, string> = {
  sum: "sum",
  min: "min",
  max: "max",
  avg: "mean",
  count: "count",
};
const AGG_OPERATORS = new Set([...Object.keys(RQL_FUNCTIONS), "aggregate"]);

function rqlLeaf(op: string, args: RqlArg[]): Expr {
  const comparator = RQL_COMPARATORS[op];
  if (comparator === undefined) {
    throw new QueryError(`Unsupported RQL operator: \`${op}\``);
  }
  const field = args[0] as string;
  let value = args[1] as unknown;
  const [family, key] = resolveField(field);
  if (comparator === "in" || comparator === "not_in") {
    value = Array.isArray(value) ? value : [value];
  } else if (
    (comparator === "like" || comparator === "ilike") &&
    typeof value === "string"
  ) {
    value = value.replace(/\*/g, "");
  }
  const lookup = comparator === "eq" ? key : `${key}__${comparator}`;
  return FAMILIES[family]({ [lookup]: value });
}

export function rqlToExpr(node: RqlNode): Expr {
  const { name, args } = node;
  if (name === "and" || name === "or") {
    const result = combine(
      args.map((a) => rqlToExpr(a as RqlNode)),
      name === "and" ? AND : OR,
    );
    if (result === null) throw new QueryError(`Empty RQL group: \`${name}\``);
    return result;
  }
  if (name === "not") {
    return rqlToExpr(args[0] as RqlNode).not();
  }
  return rqlLeaf(name, args);
}

function alephField(leaf: Leaf): string {
  return leaf.family === "P" ? `properties.${leaf.field}` : leaf.field;
}

function leafToRql(leaf: Leaf): RqlNode {
  const op = TO_RQL_OPERATORS[leaf.comparator];
  if (op === undefined) {
    throw new QueryError(
      `Comparator \`${leaf.comparator}\` is not expressible as RQL`,
    );
  }
  let value: RqlArg = leaf.value as string;
  if (op === "in" || op === "out") {
    value = (leaf.value as string[]).map((v) => String(v)).sort(byString);
  }
  return { name: op, args: [alephField(leaf), value] };
}

export function exprToRql(expr: Expr): RqlNode {
  const group = expr.connector === OR ? "or" : "and";
  const parts: RqlNode[] = [];
  for (const child of expr.children) {
    if (child instanceof Expr) {
      const childAst = exprToRql(child);
      if (!child.negated && childAst.name === group) {
        parts.push(...(childAst.args as RqlNode[]));
      } else {
        parts.push(childAst);
      }
    } else {
      parts.push(leafToRql(child));
    }
  }
  if (parts.length === 0) {
    throw new QueryError("Cannot serialize an empty query to RQL");
  }
  const body = parts.length === 1 ? parts[0] : { name: group, args: parts };
  if (expr.negated) return { name: "not", args: [body] };
  return body;
}

// --- aggregations <-> RQL ----------------------------------------------------

function metricAggs(node: RqlNode, groups: string[]): Agg[] {
  const func = RQL_FUNCTIONS[node.name];
  if (func === undefined) {
    throw new QueryError(`Unsupported RQL aggregate operator: \`${node.name}\``);
  }
  return node.args.map((prop) => new Agg(func, String(prop), groups));
}

function nodeAggs(node: RqlNode): Agg[] {
  if (node.name === "aggregate") {
    const groups = node.args.filter((a) => typeof a === "string") as string[];
    const aggs: Agg[] = [];
    for (const arg of node.args) {
      if (isNode(arg)) aggs.push(...metricAggs(arg, groups));
    }
    return aggs;
  }
  return metricAggs(node, []);
}

function aggsToRql(aggs: Agg[]): RqlNode[] {
  const ungrouped: RqlNode[] = [];
  const grouped = new Map<string, { groups: string[]; nodes: RqlNode[] }>();
  const sorted = [...aggs].sort(
    (a, b) =>
      byString(a.groups.join(","), b.groups.join(",")) ||
      byString(a.func, b.func) ||
      byString(a.prop, b.prop),
  );
  for (const agg of sorted) {
    const node: RqlNode = { name: TO_RQL_FUNCTIONS[agg.func], args: [agg.prop] };
    if (agg.groups.length) {
      const gk = agg.groups.join(",");
      if (!grouped.has(gk)) grouped.set(gk, { groups: agg.groups, nodes: [] });
      grouped.get(gk)!.nodes.push(node);
    } else {
      ungrouped.push(node);
    }
  }
  const nodes: RqlNode[] = [...ungrouped];
  for (const { groups, nodes: metrics } of grouped.values()) {
    nodes.push({ name: "aggregate", args: [...groups, ...metrics] });
  }
  return nodes;
}

/** Parse an RQL query string into a filter `Expr` and aggregation specs. */
export function parseRqlQuery(value: string): [Expr | null, Agg[]] {
  const data = parseRql(value);
  if (!data) return [null, []];
  const aggs: Agg[] = [];
  if (AGG_OPERATORS.has(data.name)) {
    return [null, nodeAggs(data)];
  }
  if (data.name === "and") {
    const filters: RqlNode[] = [];
    for (const child of data.args) {
      if (isNode(child) && AGG_OPERATORS.has(child.name)) {
        aggs.push(...nodeAggs(child));
      } else {
        filters.push(child as RqlNode);
      }
    }
    const expr = combine(
      filters.map((f) => rqlToExpr(f)),
      AND,
    );
    return [expr, aggs];
  }
  return [rqlToExpr(data), aggs];
}

/** Serialize a filter tree and aggregation specs to an RQL query string. */
export function toRql(expr: Expr | null, aggs: Agg[] = []): string {
  const nodes: RqlNode[] = [];
  if (expr && !expr.isEmpty) {
    const filterAst = exprToRql(expr);
    if (!expr.negated && filterAst.name === "and") {
      nodes.push(...(filterAst.args as RqlNode[]));
    } else {
      nodes.push(filterAst);
    }
  }
  nodes.push(...aggsToRql(aggs));
  if (nodes.length === 0) return "";
  if (nodes.length === 1) return unparseRql(nodes[0]);
  return unparseRql({ name: "and", args: nodes });
}
